File: src.kotlin/alphaTab/android/src/main/java/alphaTab/core/ecmaScript/Uint16Array.kt
package alphaTab.core.ecmaScript

@Suppress("NOTHING_TO_INLINE")
@ExperimentalUnsignedTypes
internal class Uint16Array : Iterable<UShort> {
    private val _data: UShortArray

    public val length: Double
        get() {
            return _data.size.toDouble()
        }

    public constructor(size: Double) {
        _data = UShortArray(size.toInt())
    }

    public inline operator fun get(index: Double): Double {
        return _data[index.toInt()].toDouble()
    }

    public inline operator fun set(index: Double, value: Double) {
        _data[index.toInt()] = value.toInt().toUShort()
    }

    public inline operator fun get(index: Int): Double {
        return _data[index].toDouble()
    }

    public inline operator fun set(index: Int, value: Double) {
        _data[index] = value.toUInt().toUShort()
    }

    override fun iterator(): Iterator<UShort> {
        return _data.iterator()
    }
}

File: src.csharp/AlphaTab/Core/EcmaScript/Uint16Array.cs
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace AlphaTab.Core.EcmaScript;

internal class Uint16Array : IEnumerable<ushort>
{
    private readonly ushort[] _data;

    public double Length => _data.Length;

    public Uint16Array(double size)
    {
        _data = new ushort[(int) size];
    }

    public double this[double index]
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => _data[(int) index];
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        set => _data[(int) index] = (ushort) value;
    }

    public IEnumerator<ushort> GetEnumerator()
    {
        return ((IEnumerable<ushort>)_data).GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}